keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+    - key: environment,
+      operator: Exists
+  enforcement: [Deny]
+```
+
+With this binding, the test and global policy bindings overlap. Resources
+admitted to test environment would then be checked against both policy
+configurations.
+
+To summarize, this "API Shape" separates admission policy _definition_ from
+_configuration_. This has a couple advantages:
+
+- Access to, and delegation of, policy configuration is more manageable. In
+  particular, Kubernetes RBAC works well with this design.
+- Without the separation, the next most obvious API shape would be to encode
+  everything into a single resource, which could easily become very large and
+  run into resource size limits.
+
+##### Policy Definitions
+
+Policy definitions are responsible for:
+
+- Defining what validations the policy enforces and how violations are reported 
+- Defining how a policy may be configured
+
+Each `ValidatingAdmissionPolicy` resource contains a `spec.match` to declare
+what resources it validates. This field is required.
+
+- `spec.match` constrains which resources this policy can be applied to. Policy
+  bindings each have match rules with further narrow this constraint, but cannot
+  expand it. This allows the CEL expressions to make safe assumptions. E.g. a
+  CEL expression that is constrained to CREATES and UPDATES of resources is
+  guaranteed the root `object` variable is never null, but a CEL expression that
+  might need to evaluate a DELETE must handle the root `object` variable being
+  null. See below "Match criteria" section for how match criteria is described.
+- `spec.match` guides CEL expression type checking, see the below "Type safety"
+  section for more details. 
+
+CEL expressions have access to the contents of the `AdmissionReview` type,
+organized into CEL variables as well as some other useful variables:
+
+- 'object'
+- 'oldObject'
+- 'review'
+  - 'requestResource' (GVR)
+  - 'resource' (GVR)
+  - 'name'
+  - 'namespace'
+  - 'operation'
+  - 'userInfo'
+  - 'dryRun'
+  - 'options'
+- 'config' - configuration data of the policy configuration being validated
+
+See below "Reporting violations to Clients" for more detail about how the
+`spec.validations` field works and how violations are reported.
+
+##### Policy Configuration
+
+`PolicyBinding` resources and parameter CRDs together define how cluster
+administrators configure policies for clusters.
+
+Each `PolicyBinding` contains:
+
+- `spec.policy` - A reference to the policy being configured
+- `spec.match` - Match criteria for which resources the policy should validate
+- `spec.params` - Reference to the custom resource containing the params to use
+  when validating resources 
+- `spec.enforcement` - How validation failures should be enforced
+
+Example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyBinding
+metadata:
+  name: "xyzlimit-scale.example.com"
+spec:
+  policy: xyzlimit-scale.example.com
+  params: xyzlimit-scale-settings.example.com
```

Comment:
If we're only stating the resource, is this necessary at all? Can't we get that from the policy?

Reply:
I was expecting a 1:1 relationship between a policy binding and a param CR. (and a 1:1 between policy binding and policy definition).

Reply:
I expect multiple bindings per param CR. If no other reason, consider that there will eventually be a namespaced binding if I understood phase 2 correctly.

For RBAC, without looking, I'd bet there's maybe 5x the number of bindings as roles in a typical cluster, I expect this to be similar, maybe not quite as extreme because policy & params will serve some of that need.

Reply:
I think a diagram with arrows or at least a text explanation of the relationship between these objects would be helpful.

Reply:
Oh yes, I'm sorry, I was tired when I wrote the above.

relationships should be:

- 1:N policy definition to policy binding
- 1:N param CR to policy binding

?

I was thinking of putting the references from each policy binding to one policy definition and one param CR to achieve this.

Reply:
Oh, @lavalamp were you thinking it should be {1:N definition:binding} and all params are applied to all bindings?

Reply:
I was last night, but now I think that was wrong -- I hadn't realized that a single params object can work with multiple policy objects. So you need to state both, I think.